src/Group.jl

```diff
@@ -0,0 +1,397 @@
+
+"""
+    IsMatrixGroup
+
+A `Trait` to indicate that a group's default representation is a matrix
+representation. For groups with a `MultiplicationOperation`, this indicates
+that the group is a subgroup of the General Linear Group.
+"""
+@traitdef IsMatrixGroup{G}
+
+@doc doc"""
+    inv(G::AbstractGroupManifold, x)
+
+Inverse $x^{-1}$ of an element $x$, such that
+$x \cdot x^{-1} = x^{-1} \cdot x = e$.
+"""
+function inv(G::AbstractGroupManifold, x)
+    error("inv not implemented on $(typeof(G)) for point $(typeof(x))")
+end
+
+@doc doc"""
+    identity(G::AbstractGroupManifold, x)
+
+Identity element $e$, such that for any element $x$,
+$x \cdot e = e \cdot x = x$. The returned element is of a similar type to `x`.
+"""
+function identity(G::AbstractGroupManifold, x)
+    error("identity not implemented on $(typeof(G)) for point $(typeof(x))")
+end
+
+@doc doc"""
+    left_action(G::AbstractGroupManifold, x, p)
+
+For an element $p$ of some set, compute the left action of group element $x$
+on $p$, i.e. $x \cdot p$.
+"""
+function left_action(G::AbstractGroupManifold, x, p)
+    error("left_action not implemented on $(typeof(G)) for point $(typeof(x)) and object $(typeof(p))")
+end
+
+@doc doc"""
+    right_action(G::AbstractGroupManifold, x, p)
+
+For an element $p$ of some set, compute the right action of group element $x$
+on $p$, i.e. $p \cdot x$. The default right action is constructed from the left
+action, i.e. $p \cdot x = x^{-1} \cdot p$.
+"""
+function right_action(G::AbstractGroupManifold, x, p)
+    return left_action(G, inv(G, x), p)
+end
+
+"""
+    compose(G::AbstractGroupManifold, x...)
+
+Compose elements of `G` using their left translation upon each other.
+"""
+function compose(G::AbstractGroupManifold, x...)
+    return reduce((x, y)->left_action(G, x, y), x)
+end
+
+*(G::AbstractGroupManifold, x, y) = compose(G, x, y)
+
+/(G::AbstractGroupManifold, x, y) = compose(G, x, inv(G, y))
+
+\(G::AbstractGroupManifold, x, y) = compose(G, inv(G, x), y)
+
+@doc doc"""
+    conjugate(G::AbstractGroupManifold, x, y)
+
+Compute the conjugate of the element $y$ by element $x$, i.e.
+$x \cdot y \cdot x^{-1}$
+"""
+function conjugate(G::AbstractGroupManifold, x, y)
+    return compose(G, x, y, inv(G, x))
+end
+
+function adjoint_representation(G::AbstractGroupManifold, x, v)
+    return left_action(G, left_action(G, x, v), inv(G, x))
+end
+
+function adjoint_representation_derivative(G::AbstractGroupManifold, v, w)
+    return lie_bracket(G, v, w)
+end
+
+function lie_bracket(G::AbstractGroupManifold, v, w)
+    error("lie_bracket not implemented on $(typeof(G)) for vectors $(typeof(v)) and $(typeof(w))")
+end
+
+function inner(G::AbstractGroupManifold, x, v, w)
+    return dot(left_action(G, x, v), left_action(G, x, w))
+end
+
+"""
+    exp!(G::AbstractGroupManifold, y, v)
+
+Exponential map of tangent vector `v` at the identity element of group `G`.
+Result is saved to `y`.
+"""
+function exp!(G::AbstractGroupManifold, y, v)
```

Comment:
I'm afraid adding these methods on top of what `Manifold` already has can lead to confusion and surprises. Maybe `exp_id!` would be a better name?

Reply:
Similar comment for `exp`, `log`, `log!`.

Reply:
I agree it could get messy. I'm wondering if a better approach is to define a special `Identity` type (similar to `LinearAlgebra.I`, but it's only an identity element, not a `UniformScaling`).  We can then keep `exp` and `log` but dispatch on whether the identity element is provide.

Reply:
Yes, a special identity type similar to `LinearAlgebra.I` would be even better.

Reply:
Done

Reply:
Looks great, I'm just not sure if defining a global constant `e` is a good idea. Julia already has a constant `ℯ` (`\euler`) which looks very similar.

Reply:
Sorry, never mind, there is no such constant :+1: .

Reply:
Oh, there's no global constant `e` here (though I would love a global constant; it's just that as you mentioned `I` and `ℯ` are already taken.). I just use the letter `e` for a local variables as a convention.

Reply:
Beat me to it. 👍